[LiveComponent][TwigComponent] Add support for dynamic template resolution (#3361)
symfony/ux: src/TwigComponent/doc/index.rst

        {{ message_formatter('...') }}
    </twig:Alert>

Dynamic Templates
-----------------

.. versionadded:: 2.33

    The ability to dynamically resolve templates via the ``FromMethod`` attribute was added.

Sometimes, you need to render a different template based on the component state.

> It is possible to reference a component method by passing ``FromMethod`` to the ``template`` option of ``AsTwigComponent`, which will be used to compute the template to use before rendering the component::

Reviewer: ```suggestion
It is possible to reference a component method by passing ``FromMethod`` to the ``template`` option of ``AsTwigComponent``, which will be used to compute the template to use before rendering the component::
```

    namespace App\Twig\Components;

    use Symfony\UX\TwigComponent\Attribute\AsTwigComponent;
    use Symfony\UX\TwigComponent\Attribute\FromMethod;

    #[AsTwigComponent(template: new FromMethod('getTemplate'))]
    class SearchResults

>     {
>         public string $layout = 'rows';
>         public function getTemplate(): string
>         {

Reviewer: ```suggestion
        public string $layout = 'rows';

        public function getTemplate(): string
        {
```

        {{ message_formatter('...') }}
    </twig:Alert>

Dynamic Templates
-----------------

.. versionadded:: 2.33

    The ability to dynamically resolve templates via the ``FromMethod`` attribute was added.

> Sometimes, you need to render a different template based on the component's state or props.

Reviewer: ```suggestion
Sometimes, you need to render a different template based on the component state.
```

> Instead of a static string, you can pass a ``FromMethod`` instance to the ``template`` option of the ``AsTwigComponent`` attribute. This tells Symfony to call the specified method to determine the template path at runtime::

Reviewer: ```suggestion
It is possible to reference a component method by passing ``FromMethod`` to the ``template`` option of ``AsTwigComponent`, which will be used to compute the template to use before rendering the component::
```

    // src/Twig/Components/UserProfile.php
    namespace App\Twig\Components;

    use Symfony\UX\TwigComponent\Attribute\AsTwigComponent;
    use Symfony\UX\TwigComponent\Attribute\FromMethod;

    #[AsTwigComponent('user_profile', template: new FromMethod('getTemplateForRole'))]
    class UserProfile
    {
        public function getTemplateForRole(): string
        {
            if ($this->isGranted('ROLE_ADMIN')) {
                return 'components/user_profile/admin_dashboard.html.twig';
            }

            return 'components/user_profile/public_card.html.twig';
        }

>     }

Reviewer: Please re-use example from https://github.com/symfony/ux/pull/3361#discussion_r2880911533

    </twig:Alert>

.. versionadded:: 2.33

    The ability to dynamically resolve templates via the ``getTemplate()`` method was added in TwigComponents 2.x.

Sometimes, you need to reuse the exact same component logic (PHP class) but render a completely different template based on the component's state or props (e.g., rendering a desktop vs. mobile version of a search dropdown or a navigation menu).

Instead of duplicating the component or writing a ``PreRenderEvent`` listener, you can define a ``getTemplate()`` method in your component class. If this public method exists, its return value will be used as the template path, overriding the default path or the one defined in the ``AsTwigComponent`` attribute::

    // src/Twig/Components/SearchForm.php
    namespace App\Twig\Components;

    use Symfony\UX\TwigComponent\Attribute\AsTwigComponent;

    #[AsTwigComponent]
    class SearchForm
    {
        public string $theme = 'desktop';

        public function getTemplate(): string
        {
            return 'mobile' === $this->theme
                ? 'components/SearchForm/mobile.html.twig'
                : 'components/SearchForm/desktop.html.twig';

>         }

Reviewer: This is technically a BC break for users that already have a method `getTemplate` in their component.

        {{ message_formatter('...') }}
    </twig:Alert>

Dynamic Templates
-----------------

.. versionadded:: 2.33

    The ability to dynamically resolve templates via the ``FromMethod`` attribute was added.

Sometimes, you need to render a different template based on the component's state or props (e.g., a desktop vs. mobile version of a search dropdown).

Instead of a static string, you can pass a ``FromMethod`` instance to the ``template`` option of the ``AsTwigComponent`` attribute. This tells Symfony to call the specified method to determine the template path at runtime::

    // src/Twig/Components/SearchForm.php
    namespace App\Twig\Components;

    use Symfony\UX\TwigComponent\Attribute\AsTwigComponent;
    use Symfony\UX\TwigComponent\Attribute\FromMethod;

    #[AsTwigComponent(template: new FromMethod('getDynamicTemplate'))]
    class SearchForm
    {
        public string $theme = 'desktop';

        public function getDynamicTemplate(): string
        {
            return 'mobile' === $this->theme
                ? 'components/SearchForm/mobile.html.twig'
                : 'components/SearchForm/desktop.html.twig';
        }

>     }

Reviewer: I feel like this example could be more realistic, usually mobile-vs-desktop happens in the browser through CSS classes:
```suggestion
    #[AsTwigComponent(template: new FromMethod('getTemplate'))]
    class SearchResults
    {
        public string $layout = 'rows';

        public function getTemplate(): string
        {
            return 'rows' === $this->layout
                ? 'components/SearchResults/rows.html.twig'
                : 'components/SearchResults/grid.html.twig';
        }
    }
```

Now, you can switch the template by passing the ``theme`` prop:

.. code-block:: html+twig

    {# Renders components/SearchForm/mobile.html.twig #}
    <twig:SearchForm theme="mobile" />

    {# Renders components/SearchForm/desktop.html.twig #}

>     <twig:SearchForm theme="desktop" />

Reviewer: (to adapt given the previous comment)
